chain/trie/pruning.go

```diff
@@ -0,0 +1,305 @@
+package trie
+
+import (
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/nknorg/nkn/chain/db"
+	"github.com/nknorg/nkn/common"
+)
+
+type RefCounts struct {
+	trie   *Trie
+	counts map[common.Uint256]uint32
+
+	startRefCountHeight  uint32
+	startPruningHeight   uint32
+	targetRefCountHeight uint32
+	targetPruningHeight  uint32
+
+	dbCounter   int32
+	nodeCounter int32
+}
+
+func NewRefCounts(t *Trie, refCountTargetHeight, pruningTargetHeight uint32) *RefCounts {
+	ref := &RefCounts{
+		counts:               make(map[common.Uint256]uint32, 0),
+		trie:                 t,
+		targetRefCountHeight: refCountTargetHeight,
+		targetPruningHeight:  pruningTargetHeight,
+	}
+
+	heightBuffer, err := t.db.Get(db.TriePrunedHeightKey())
+	if err != nil {
+		ref.startPruningHeight = 0
+	} else {
+		ref.startPruningHeight = binary.LittleEndian.Uint32(heightBuffer) + 1
+	}
+
+	heightBuffer, err = t.db.Get(db.TrieRefCountHeightKey())
+	if err != nil {
+		ref.startRefCountHeight = 0
+	} else {
+		ref.startRefCountHeight = binary.LittleEndian.Uint32(heightBuffer) + 1
+	}
+
+	return ref
+}
+
+func (ref *RefCounts) RebuildRefCount() error {
+	iter := ref.trie.db.NewIterator([]byte{byte(db.TRIE_RefCount)})
+	for iter.Next() {
+		hs, _ := common.Uint256ParseFromBytes(iter.Key()[1:])
+		height := binary.LittleEndian.Uint32(iter.Value())
+		ref.counts[hs] = height
+	}
+	iter.Release()
+
+	return nil
+}
+
+func (ref *RefCounts) DumpInfos(offset uint32, pruning bool) error {
+	if !pruning {
+		fmt.Println("refcount:", time.Now().Format("2006-01-02 15:04:05"), ref.startRefCountHeight+offset, ref.dbCounter, ref.nodeCounter, len(ref.counts))
+	} else {
+		fmt.Println("pruning: ", time.Now().Format("2006-01-02 15:04:05"), ref.startPruningHeight+offset, ref.dbCounter, ref.nodeCounter, len(ref.counts))
+	}
+
+	return nil
+}
+
+func (ref *RefCounts) CreateRefCounts(hash common.Uint256) error {
+	root, err := ref.trie.resolveHash(hash.ToArray(), true)
+	if err != nil {
+		return err
+	}
+
+	ref.trie.root = root
+	ref.dbCounter = 0
+	ref.nodeCounter = 0
+	err = ref.createRefCounts(ref.trie.root)
+	if err != nil {
+		panic(err)
+	}
+
+	return nil
+}
+
+func (ref *RefCounts) createRefCounts(n node) error {
+	ref.nodeCounter++
+	switch n := n.(type) {
+	case *shortNode:
+		hash, _ := n.cache()
+		hs, _ := common.Uint256ParseFromBytes(hash)
+		_, ok := ref.counts[hs]
+		if !ok {
```

Comment:
I think it's safer to use `ref.counts[hs] == 0` just in case we store 0 in count